// params-verifier/lib/BooleanValidator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _Err = require('./Err');

var _ValidatorBase2 = require('./ValidatorBase');

var _ValidatorBase3 = _interopRequireDefault(_ValidatorBase2);

var _dataType = require('./dataType');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BooleanValidator = function (_ValidatorBase) {
    (0, _inherits3.default)(BooleanValidator, _ValidatorBase);

    function BooleanValidator(data, options) {
        (0, _classCallCheck3.default)(this, BooleanValidator);

        var _this = (0, _possibleConstructorReturn3.default)(this, (BooleanValidator.__proto__ || (0, _getPrototypeOf2.default)(BooleanValidator)).call(this));

        _this.type = 'boolean';
        _this.value = data;
        _this.options = options || {};
        return _this;
    }

    (0, _createClass3.default)(BooleanValidator, [{
        key: 'singleField',
        value: function singleField(name, options) {
            this.currentFieldInfo = this.fetchFieldInfo(name, this.value, this.type, options);
            this.validate();
        }
    }, {
        key: 'checkOptionsParamsSupport',
        value: function checkOptionsParamsSupport() {
            this.checkOptionsRequired();
        }
    }, {
        key: 'startValidateProcedure',
        value: function startValidateProcedure() {
            this.tryCastType();
            this.verifyType();
        }
    }, {
        key: 'tryCastType',
        value: function tryCastType() {
            var value = this.currentFieldInfo.value;


            if (typeof value === 'string' && ['true', 'false'].indexOf(value) > -1) {
                this.currentFieldInfo.value = value === 'true';
            }
        }
    }, {
        key: 'verifyType',
        value: function verifyType() {
            var _currentFieldInfo = this.currentFieldInfo,
                value = _currentFieldInfo.value,
                name = _currentFieldInfo.name;


            if (!(0, _dataType.isBool)(value)) {
                throw new _Err.ParamsError('the type of ' + name + ' is not boolean.');
            }
        }
    }]);
    return BooleanValidator;
}(_ValidatorBase3.default);

exports.default = BooleanValidator;
