// params-verifier/lib/dataType.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _isNan = require('babel-runtime/core-js/number/is-nan');

var _isNan2 = _interopRequireDefault(_isNan);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

exports.isObj = isObj;
exports.isArr = isArr;
exports.isFunc = isFunc;
exports.isDat = isDat;
exports.isReg = isReg;
exports.isBool = isBool;
exports.isNum = isNum;
exports.isStr = isStr;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isObj(data) {
    return (typeof data === 'undefined' ? 'undefined' : (0, _typeof3.default)(data)) === 'object' && Object.prototype.toString.call(data) === '[object Object]';
}

function isArr(data) {
    return Array.isArray(data);
}

function isFunc(data) {
    return Object.prototype.toString.call(data) === '[object Function]';
}

function isDat(data) {
    return (typeof data === 'undefined' ? 'undefined' : (0, _typeof3.default)(data)) === 'object' && Object.prototype.toString.call(data) === '[object Date]' && !(0, _isNan2.default)(data.getTime());
}

function isReg(data) {
    return Object.prototype.toString.call(data) === '[object RegExp]';
}

function isBool(data) {
    return typeof data === 'boolean';
}

function isNum(data) {
    return typeof data === 'number' && !(0, _isNan2.default)(data);
}

function isStr(data) {
    return typeof data === 'string';
}
