// params-verifier/lib/ObjectValidator.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _Err = require('./Err');

var _ValidatorBase2 = require('./ValidatorBase');

var _ValidatorBase3 = _interopRequireDefault(_ValidatorBase2);

var _selectValidator = require('./selectValidator');

var _selectValidator2 = _interopRequireDefault(_selectValidator);

var _dataType = require('./dataType');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ObjectValidator = function (_ValidatorBase) {
    (0, _inherits3.default)(ObjectValidator, _ValidatorBase);

    function ObjectValidator(data, options) {
        (0, _classCallCheck3.default)(this, ObjectValidator);

        var _this = (0, _possibleConstructorReturn3.default)(this, (ObjectValidator.__proto__ || (0, _getPrototypeOf2.default)(ObjectValidator)).call(this));

        _this.value = data;
        _this.options = options || {};
        _this.type = 'object';
        _this.fieldsList = {};
        return _this;
    }

    (0, _createClass3.default)(ObjectValidator, [{
        key: 'singleField',
        value: function singleField(name, options) {
            this.currentFieldInfo = this.fetchFieldInfo(name, this.value, this.type, options);
            this.validate();
        }
    }, {
        key: 'field',
        value: function field(name, type, options) {
            var typeValidator = (0, _selectValidator2.default)(type)(this.value[name], options);
            typeValidator.currentFieldInfo = this.fetchFieldInfo(name, typeValidator.value, typeValidator.type, options);
            var isFiltered = typeValidator.validate();
            if (isFiltered) {
                this.fieldsList[name] = isFiltered[name];
            }
            return this;
        }
    }, {
        key: 'checkOptionsParamsSupport',
        value: function checkOptionsParamsSupport() {
            this.checkOptionsValidator();
            this.checkOptionsRequired();
        }
    }, {
        key: 'startValidateProcedure',
        value: function startValidateProcedure() {
            this.verifyType();
            this.runValidator();
        }
    }, {
        key: 'verifyType',
        value: function verifyType() {
            var _currentFieldInfo = this.currentFieldInfo,
                value = _currentFieldInfo.value,
                name = _currentFieldInfo.name;

            if (!(0, _dataType.isObj)(value)) {
                throw new _Err.ParamsError('the type of ' + name + ' is not object.');
            }
        }
    }, {
        key: 'filter',
        value: function filter() {
            return this.fieldsList;
        }
    }]);
    return ObjectValidator;
}(_ValidatorBase3.default);

exports.default = ObjectValidator;
module.exports = exports['default'];
